lib/streamlit/runtime/pages_manager.py

```diff
@@ -0,0 +1,341 @@
+        self,
+        callback: Callable[[str], None],
+    ) -> Callable[[], None]:
+        # V2 strategy does not handle any pages changed event
+        return lambda: None
+
+
+class PagesManager:
+    """
+    PagesManager is responsible for managing the set of pages based on the
+    strategy. By default, PagesManager uses V1 which relies on the original
+    assumption that there exists a `pages` directory with all the scripts.
+
+    If the `pages` are being set directly, the strategy is switched to V2.
+    This indicates someone has written an `st.navigation` call in their app
+    which informs us of the pages.
+
+    NOTE: Each strategy handles its own thread safety when accessing the pages
+    """
+
+    DefaultStrategy: Type[PagesStrategyV1 | PagesStrategyV2] = PagesStrategyV1
+
+    def __init__(self, main_script_path, script_cache=None, **kwargs):
+        self._main_script_path: ScriptPath = main_script_path
+        self._main_script_hash: PageHash = calc_md5(main_script_path)
+        self._current_page_hash: PageHash = self._main_script_hash
+        self.pages_strategy = PagesManager.DefaultStrategy(self, **kwargs)
+        self._script_cache: ScriptCache | None = script_cache
+        self._intended_page_script_hash: PageHash | None = None
+        self._intended_page_name: PageName | None = None
+
+    @property
+    def current_page_hash(self) -> PageHash:
+        return self._current_page_hash
+
+    @property
+    def main_script_path(self) -> ScriptPath:
+        return self._main_script_path
+
+    @property
+    def main_script_hash(self) -> PageHash:
+        return self._main_script_hash
+
+    @property
+    def intended_page_name(self) -> PageName | None:
+        return self._intended_page_name
+
+    @property
+    def intended_page_script_hash(self) -> PageHash | None:
+        return self._intended_page_script_hash
+
+    def get_main_page(self) -> PageInfo:
+        return {
+            "script_path": self._main_script_path,
+            "page_script_hash": self._main_script_hash,
+        }
+
+    def get_current_page_script_hash(self) -> PageHash:
+        return self._current_page_hash
+
+    def set_current_page_script_hash(self, page_hash: PageHash) -> None:
+        self._current_page_hash = page_hash
+
+    def get_active_script_hash(self) -> PageHash:
+        return self.pages_strategy.get_active_script_hash()
+
+    def set_active_script_hash(self, page_hash: PageHash):
+        return self.pages_strategy.set_active_script_hash(page_hash)
+
+    def set_script_intent(
+        self, page_script_hash: PageHash, page_name: PageName
+    ) -> None:
+        self._intended_page_script_hash = page_script_hash
+        self._intended_page_name = page_name
+
+    def get_initial_active_script(
+        self, page_script_hash: PageHash, page_name: PageName
+    ) -> Optional[PageInfo]:
+        return self.pages_strategy.get_initial_active_script(
+            page_script_hash, page_name
```

Comment:
nit: maybe it a quick docstring to those methods. e.g. just from looking at it its not super obvious what the difference of `get_active_script_hash` and `get_current_page_script_hash` is.